enhancements/project-request-limit.md

```diff
@@ -0,0 +1,293 @@
+```
+https://{server}/apis/requestlimit.project.openshift.io/v1/requests
+```
+
+The webhook will implement `localhost` serving, as follows:
+* The webhook server should be hosted on the same node as the `kube-apiserver`.
+* The webhook server should be reachable to the `kube-apiserver` over `localhost` of the `hostnetwork`.
+* Communication between `kube-apiserver` and the webhook server must be secure using `x509`. 
+* The webhook server should only allow connection over `localhost`, thus it would disallow connection(s) originating 
+  from outside the host.
+
+This is how we are going to deploy the webhook server.
+* Use `DaemonSet` to deploy the pods of the webhook server. 
+* Enable `hostnetwork` for `PodSpec` - `hostnetwork: true`.
+* Select a host port that is available, `9410`. 
+* Apply appropriate `nodeSelector` and `tolerations` to the `PodSpec` so that the `Pods` are scheduled on to the master
+ node(s). This way `kube-apiserver` can access the webhook server using `localhost`.
+* The webhook generates `localhost` serving certs. 
+* The webhook binds to `127.0.0.1`.
+* `kube-apiserver` uses `https://localhost` to access the admission webhook server. `clientConfig.URL` of the 
+`ValidatingWebhookConfiguration` should be set to `https://localhost:{host port}/apis/{api path}`.
+
+```
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: MutatingWebhookConfiguration
+webhooks:
+- clientConfig:
+    url: https://localhost:9410/apis/requestlimit.project.openshift.io/v1/requests
+```
+
+#### The Operator
+The operator will allow a cluster admin to:
+* Enable or disable `ProjectRequestLimit` admission webhook on a cluster.
+* Specify `limits` for `Project` requests that the `ProjectRequestLimit` admission webhook can enforce.
+* Manage other aspects of the operand lifecycle including cert rotation.
+
+The operator will define a CRD to expose its API. The cluster admin will interact with the operator via a custom resource:
+* We treat the `ProjectRequestLimit` admission webhook as a cluster singleton.
+* The CRD will be `cluster-scoped`.
+* The operator will be reconciling a CR named `cluster`, it will ignore other custom resources.
+
+```
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  name: projectrequestlimitconfigs.requestlimit.project.openshift.io
+spec:
+  scope: Cluster
+```  
+
+This is how a cluster-admin can enable the `ProjectRequestLimit` admission webhook on a cluster with some `limits` to be enforced. 
+```
+apiVersion: requestlimit.project.openshift.io/v1
+kind: ProjectRequestLimitConfig
+metadata:
+  name: cluster
+spec:
+  limits:
+  - selector:
+      level: admin
+  - selector:
+      level: advanced
+    maxProjects: 10
+  - maxProjects: 2
+```
+
+The operator will ensure that:
+* Changes in configuration are propagated to the `ProjectRequestLimit` admission webhook and the new configuration should 
+  take effect immediately.
```

Comment:
what does this mean? Does the webhook (operand) listen on this CR?

Reply:
No, the operand is immutable and will not manage any aspect of its own life cycle. We have an operator to manage it, so the operator will ensure that the webhook reloads the configuration.